src/cask/scheduler/WorkStealingScheduler.cpp

```diff
@@ -0,0 +1,138 @@
+//          Copyright Tango Tango, Inc. 2020 - 2021.
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          https://www.boost.org/LICENSE_1_0.txt)
+
+#include "cask/scheduler/WorkStealingScheduler.hpp"
+#include <cstdlib>
+#include <cassert>
+
+namespace cask::scheduler {
+
+WorkStealingScheduler::WorkStealingScheduler(unsigned int poolSize, std::optional<int> priority)
+    : running_thread_count(0)
+{
+    assert(poolSize > 0 && "Pool size must be greater than 0");
+    auto idle_callback = std::bind(&WorkStealingScheduler::onThreadIdle, this->weak_from_this());
+    auto resume_callback = std::bind(&WorkStealingScheduler::onThreadResume, this->weak_from_this());
+    auto request_work_callback = std::bind(&WorkStealingScheduler::onThreadRequestWork, this->weak_from_this());
+
+    for(unsigned int i = 0; i < poolSize; i++) {
+        auto sched = std::make_shared<SingleThreadScheduler>(priority, std::nullopt, idle_callback, resume_callback, request_work_callback);
+        auto thread_id = sched->get_run_thread_id();
+
+        thread_ids.push_back(thread_id);
+        schedulers[thread_id] = sched;
+    }
+}
+
+void WorkStealingScheduler::submit(const std::function<void()>& task) {
+    auto thread_handle = std::this_thread::get_id();
+    auto scheduler_lookup = schedulers.find(thread_handle);
+
+    if (scheduler_lookup != schedulers.end()) {
+        scheduler_lookup->second->submit(task);
+    } else {
+        auto random_index = std::rand() % thread_ids.size();
+        auto random_thread_id = thread_ids[random_index];
+        schedulers[random_thread_id]->submit(task);
+    }
+}
+
+void WorkStealingScheduler::submitBulk(const std::vector<std::function<void()>>& tasks) {
+    auto thread_handle = std::this_thread::get_id();
+    auto scheduler_lookup = schedulers.find(thread_handle);
+
+    if (scheduler_lookup != schedulers.end()) {
+        scheduler_lookup->second->submitBulk(tasks);
+    } else {
+        auto random_index = std::rand() % thread_ids.size();
+        auto random_thread_id = thread_ids[random_index];
+        schedulers[random_thread_id]->submitBulk(tasks);
+    }
+}
+
+CancelableRef WorkStealingScheduler::submitAfter(int64_t milliseconds, const std::function<void()>& task) {
+    auto thread_handle = std::this_thread::get_id();
+    auto scheduler_lookup = schedulers.find(thread_handle);
+
+    if (scheduler_lookup != schedulers.end()) {
+        return scheduler_lookup->second->submitAfter(milliseconds, task);
+    } else {
+        auto random_index = std::rand() % thread_ids.size();
+        auto random_thread_id = thread_ids[random_index];
+        return schedulers[random_thread_id]->submitAfter(milliseconds, task);
+    }
+}
+
+bool WorkStealingScheduler::isIdle() const {
```

Comment:
Idle detection should use the `running_thread_count` rather than iterating and asking each thread if its idle